File: packages/shared/src/types/calculation.ts
export interface Calculation {
  id: number;
  bond_id: number;
  calc_date: string;
  dirty_price: string;
  accrued_interest: string;
  yield_to_maturity: string;
  spread: string | null;
  modified_duration: string | null;
  macaulay_duration: string | null;
  created_at: string;
}

export interface CalculationSummary {
  isin: string;
  settlement_date: string;
  clean_price: string;
  dirty_price: string;
  accrued_interest: string;
  yield_to_maturity: string;
  spread: string | null;
  macaulay_duration: string;
  modified_duration: string;
  cash_flows: { date: string; amount: string }[];
}

File: apps/api/app/schemas/calculation.py
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class CalculationRequest(BaseModel):
    bond_id: int
    calc_date: date | None = None


class CalculationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    bond_id: int
    calc_date: date
    dirty_price: Decimal
    accrued_interest: Decimal
    yield_to_maturity: Decimal
    spread: Decimal | None
    modified_duration: Decimal | None
    macaulay_duration: Decimal | None
    created_at: datetime


class CalculationSummary(BaseModel):
    isin_code: str
    bond_type: str
    maturity_date: date
    clean_price: Decimal
    dirty_price: Decimal
    accrued_interest: Decimal
    ytm: Decimal
    spread: Decimal | None
